### PR TITLE
Implement and benchmark a torch.optim.multi_tensor.adagrad implementation

### PR DESCRIPTION
**Pull request by esantorella**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59155**

Differential Revision: [D29525213](https://our.internmc.facebook.com/intern/diff/D29525213)

### CODE REVIEW

**Comment by esantorella** [comment]
@esantorella has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28774053).

**Comment by ramvenkat98** [comment]
Moving some notes about this change from the Phabricator diff to this pull request so it can be reviewed here on Github more easily:

Summary:
[x] Implement a `torch.optim._multi_tensor` version of Adagrad in `torch.optim._multi_tensor.adagrad` (compare to naive Adagrad in `torch.optim.adagrad` or other `_multi_tensor` optimizers like `torch.optim._multi_tensor.sgd`). Falls back to `torch.optim.Adagrad` logic when there is a sparse tensor in the list.
[x] Add tests for `torch.optim._multi_tensor.Adagrad` in
`test/test_optim.py` to confirm that parameters do not worsen after one step in both the dense and sparse cases
[x] Benchmark against both naive Adagrad and `apex.FusedAdagrad` in a notebook:
https://nbviewer.jupyter.org/gist/ramvenkat98/8912a8a0854d2f116754c35a70448919
[x] Confirm that this optimizer is taking the same steps as naive Adagrad

Test Plan:
[x] Unit tests in test.test_optim.py
[x] Benchmark in a notebook against both torch.optim.Adagrad and apex.FusedAdagrad
[x] Check (with the unit test in test_multi_tensor_optimizers) that the steps by both naive and multi-tensor Adagrad are the same.

Other Notes:

1. Benchmarking: We see that this implementation of Adagrad performs better than the naive implementation of Adagrad when dealing with tensors that are not too large (up to around N=3000, from the figure in the notebook); this trend is similar to the respective implementations of SGD as well, and is even observed when comparing _add vs _foreach_add (figures for these are also in the notebook). It seems that for very large tensors, the additional work performed in foreach offsets the gains from reducing the number of kernels created.
2. Unit tests: When run with a GPU, the change to test_multi_tensor_optimizers in the original pull request verifies that this optimizer takes the same steps as regular Adagrad from a known starting state. Changes to test_adagrad and test_adagrad_sparse in the existing pull request apply the tests for regular Adagrad to this version.
3. A minor thing I was confused about was the reason behind the overall factor of 5 in the formula for bandwidth in the original notebook I based my benchmarking on (https://nbviewer.jupyter.org/gist/ajtulloch/94d4a12a5dbe73e95ffc4702ea2d4fdc#); I guess it suggests that, for each parameter, there are 5 associated 4-byte variables that are transferred to/from the gpu but I'm not sure what these are (I can see three such variables - params_with_grads, grads, and state_sums). I did not worry about this since it's an overall constant factor, and it doesn't affect the relative comparison between the optimizers' performance.

File: torch/optim/_multi_tensor/adagrad.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                if has_sparse_grad:
+                    raise RuntimeError("weight_decay option is not compatible with sparse gradients")
+                torch._foreach_add_(grads, params_with_grad, alpha=weight_decay)
+
+            minus_clr = [-lr / (1 + (step - 1) * lr_decay) for step in state_steps]
+
+            if has_sparse_grad:
+                # sparse is not supported by multi_tensor. Fall back to optim.adagrad
+                # implementation.
+                F.adagrad(params_with_grad,
```
Comment (iramazanli) [inline torch/optim/_multi_tensor/adagrad.py:99]: We don't call single tensor functional optimizer from multi tensor API yet. Do you think you can do it without calling this function ? I can suggest you two ways:

1. We do have multi tensor functional optimizers:  https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/_functional.py , you can add multi-tensor implementation here and call this new function.

2. You can avoid add your code here yourself, without calling this function.

Comment (iramazanli) [inline torch/optim/_multi_tensor/adagrad.py:99]: We don't call single tensor functional optimizer from multi tensor API yet. Do you think you can do it without calling this function ? I can suggest you two ways:

1.  We do have multi tensor functional optimizers: https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/_functional.py , you can add multi-tensor implementation here and call this new function.

2. You can avoid add your code here yourself, without calling this function.

**Review by iramazanli — COMMENTED** [review]
Overall code looks good, but minor change needed regarding calling single tensor functional Adagrad.

File: torch/optim/_multi_tensor/adagrad.py (hunk 1 above)
Comment (ramvenkat98) [reply torch/optim/_multi_tensor/adagrad.py]: I've shifted the relevant section of the code to the _functional file, and used the implementation of regular optim.adagrad for sparse gradients, without calling it explicitly

**Review by iramazanli — APPROVED** [review]
Thanks for adding functional multi tensor API for Adagrad! and LGTM

**Comment by ramvenkat98** [comment]
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29525213).
